warp/Network/Wai/Handler/Warp/Run.hs

```diff
@@ -64,7 +64,7 @@ socketConnection s = do
         connSendMany = Sock.sendMany s
       , connSendAll = sendall
       , connSendFile = sendFile s writeBuf bufferSize sendall
-      , connClose = close s
+      , connClose = shutdown s ShutdownSend `E.catch` \(SomeException _) -> return ()
```

Comment:
1. Why is it sufficient to call `shutdown` but never `close` afterwards?
2. What if `shutdown()` returns `ENOTCONN`, do we leak the socket?
3. Is the catch-all exception handler good, or should it be more specific, so that we can observe bugs in the code? For example, should it not be at max some `IOError`? Should it perhaps be even specific to some `errno` we can expect, like `ENOTCONN`?
4. Of the 4 steps I list in https://github.com/nodejs/node/issues/12339#issuecomment-368124078, this does only 2 steps. Don't we have to
    * Keep `read()`ing from the socket until the empty string is read
    * Call `close()` on the socket
5. What about the case of misbehaving clients that never close their side? Will they be able to drain resources from the server by keeping ports open? Should there be some timeout/setting/mechanism to `close()` if the other side doesn't `close()` after some expected duration?

Even if these are fine, I think it deverses a code comment why they are fine.